packages/fx-core/src/component/driver/aad/update.ts

```diff
@@ -100,6 +102,19 @@ export class UpdateAadAppDriver implements StepDriver {
         };
       }
       if (axios.isAxiosError(error)) {
+        if (
+          error.response!.status == 400 &&
+          error.response!.data.error.code === aadErrorCode.permissionErrorCode
```

Comment:
Can you move the added logic to `AadAppClient.updateAadApp()` function and throw a UserError? It could be a common logic.

Should be some thing like this:
```
if (error.response.data.error.code === "CannotDeleteOrUpdateEnabledEntitlement") {
    throw new DeleteOrUpdatePermissionFailedError(...)
}
```

And the error message could be following if you don't want to include the raw http error response in the message: `Failed to update or delete an existing permission when it's enabled. One possible reason is the ACCESS_AS_USER_PERMISSION_ID environment variable is changed for selected environment. Ensure your permission id(s) are identical with the actual AAD application and try again.`

Reply:
what's the actionName in UserError Type? aadApp/update?
I thought the `AadAppClient` class was just an agent proxy, and the content processing of the response was more at the business level which in aad/update action. And it was a user error.

Reply:
The `actionName` parameter is used as error's source. We can just set `AadAppClient` as error source, or leave it empty.

In my opinion, the `AadAppClient` serves for our own business requirement, instead of trying to be a common client that will be used outside of TTK. So it can have common logics to reduce duplicate logic. For example, it will just return a client secret string instead of returning the whole password object because all our business logic just need a client secret string. `CannotDeleteOrUpdateEnabledEntitlement` error is the same: if there's no requirement to treat it in different ways, we can put it to `AadAppClient`. And we always have chance to move it outside.

Reply:
if the `AadAppClient` handle `CannotDeleteOrUpdateEnabledEntitlement` exception in `updateAadApp`, the code will updated as 
``` 
public async updateAadApp(manifest: AADManifest): Promise<AxiosResponse<any>> {
    try {
      const response = await this.axios.patch(`applications/${objectId}`, requestBody, {"axios-retry"})
      return response;
    } catch(err) {
      // handle CannotDeleteOrUpdateEnabledEntitlement
      throw error
   }
```
The above error handling method is duplicated with the `axios.isAxiosError` handling `4XX` in `UpdateAadAppDriver.execute`, shall we move all the axios.isAxiosError to AadAppClient?

Reply:
I'm ok to this, may need to consider how to reuse the same error handling logic between different functions in `AadAppClient`. We can use another PR to improve it if you have time.